File: quickstart_guides/linked_lists/java/MergeKSorted.java
/*
 *   Title: Merge K sorted list
 *
 *   Problem:
 *      Merge k sorted linked lists and return it as one sorted list. Analyze
 *      and describe its complexity.
 *
 *   Execution: javac MergeKSorted.java && java MergeKSorted
 */
import java.util.*;


public class MergeKSorted {
    public static class ListNode {
        int val;
        ListNode next;
        ListNode(int x) { val = x; }
    }
    public static ListNode mergeKLists(ListNode[] lists) {
        if (lists==null || lists.length==0) return null;

        PriorityQueue<ListNode> queue= new PriorityQueue<ListNode>(lists.length, (a,b)-> a.val-b.val);

        ListNode dummy = new ListNode(0);
        ListNode tail=dummy;

        for (ListNode node:lists)
            if (node!=null)
                queue.add(node);

        while (!queue.isEmpty()){
            tail.next=queue.poll();
            tail=tail.next;

            if (tail.next!=null)
                queue.add(tail.next);
        }
        return dummy.next;
    }
    public static void main(String[] args) {

        ListNode test_1 = new ListNode(1);
        test_1.next = new ListNode(4);
        test_1.next.next = new ListNode(5);

        ListNode test_2 = new ListNode(1);
        test_2.next = new ListNode(3);
        test_2.next.next = new ListNode(4);

        ListNode test_3 = new ListNode(2);
        test_3.next = new ListNode(6);

        ListNode[] test_input = {test_1, test_2, test_3};

        System.out.println(mergeKLists(test_input));

        System.out.println("Passed all test cases");
    }

}

File: quickstart_guides/linked_lists/python/merge_k_sorted.py
"""
Title: Merge k lists.

Problem:
    Merge k sorted linked lists and return it as one sorted list. Analyze and
    describe its complexity.

Execution: python merge_k_lists.py
"""
import unittest
from typing import List


class ListNode:
    """Class for Linked List node."""

    def __init__(self, x):
        self.val = x
        self.next = None


def merge_k_lists(lists: List[ListNode]) -> ListNode:
    """Merge K-lists."""
    nodes = []
    head = point = ListNode(0)
    for l in lists:
        while l:
            nodes.append(l.val)
            l = l.next
    for x in sorted(nodes):
        point.next = ListNode(x)
        point = point.next
    return head.next


def check_list_equal(list_1: ListNode, list_2: ListNode) -> bool:
    """Check if two lists are equal."""
    while list_1.next:
        if list_1.val != list_2.val:
            return False
        list_1 = list_1.next
        list_2 = list_2.next
    return True


class TestMergeKLists(unittest.TestCase):
    """Unit test for merge_k_lists."""

    def test_1(self):
        list_1 = ListNode(1)
        list_1.next = ListNode(4)
        list_1.next.next = ListNode(5)

        list_2 = ListNode(1)
        list_2.next = ListNode(3)
        list_2.next.next = ListNode(4)

        list_3 = ListNode(2)
        list_3.next = ListNode(6)

        k_lists = [list_1, list_2, list_3]

        # Expected output for the above test.
        expected_out = ListNode(1)
        expected_out.next = ListNode(1)
        expected_out.next.next = ListNode(2)
        expected_out.next.next.next = ListNode(3)
        expected_out.next.next.next.next = ListNode(4)
        expected_out.next.next.next.next.next = ListNode(4)
        expected_out.next.next.next.next.next.next = ListNode(5)
        expected_out.next.next.next.next.next.next.next = ListNode(6)

        out_list = merge_k_lists(k_lists)
        self.assertEqual(check_list_equal(out_list, expected_out), True)


if __name__ == "__main__":
    unittest.main()
